### PR TITLE
Version 0.1.0 finish

### PR DESCRIPTION
**Pull request by hackerchai**
1. Implement basic features and fully support with `async`.
2. Static SQL query to check error while compiling.
3. Use `ENV` variable as default database connection configuration.
4. Accomplish `README.md` and CI integration.

### CODE REVIEW

File: src/adapter.rs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+//pub const TABLE_NAME: &str = "casbin_rules";
+
+impl<'a> SqlxAdapter {
+    pub async fn new() -> Result<Self> {
+        let database_url = std::env::var("DATABASE_URL").expect("DATABASE_URL is required");
+        let pool_size: u32 = std::env::var("POOL_SIZE")
+            .ok()
+            .and_then(|s| s.parse().ok())
```
Comment (0x8f701) [inline src/adapter.rs:20]: this can be simplified. We don't need the two `ok`

File: src/adapter.rs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            .unwrap_or(32);
+
+        let pool = Pool::builder()
+            .max_size(pool_size)
+            .build(database_url.as_str())
```
Comment (0x8f701) [inline src/adapter.rs]: `&` is better than `as_str`

File: src/adapter.rs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+    async fn save_policy(&mut self, m: &mut dyn Model) -> Result<()> {
+        let mut rules = vec![];
+
+        if let Some(ast_map) = m.get_model().get("p") {
+            for (ptype, ast) in ast_map {
+                let new_rules = ast
+                    .get_policy()
+                    .into_iter()
+                    .filter_map(|x: &Vec<String>| self.save_policy_line(ptype, x));
```
Comment (0x8f701) [inline src/adapter.rs]: can this type definition be removed? because normally it will be inferred

File: src/adapter.rs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                rules.extend(new_rules);
+            }
+        }
+
+        if let Some(ast_map) = m.get_model().get("g") {
+            for (ptype, ast) in ast_map {
+                let new_rules = ast
+                    .get_policy()
+                    .into_iter()
+                    .filter_map(|x: &Vec<String>| self.save_policy_line(ptype, x));
```
Comment (0x8f701) [inline src/adapter.rs]: same here

File: src/adapter.rs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+    async fn add_policies(
+        &mut self,
+        _sec: &str,
+        ptype: &str,
+        rules: Vec<Vec<String>>,
+    ) -> Result<bool> {
+        let new_rules = rules
+            .iter()
+            .filter_map(|x: &Vec<String>| self.save_policy_line(ptype, x))
```
Comment (0x8f701) [inline src/adapter.rs]: same here

File: src/actions.rs [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+            field_values[4],
+            field_values[5]
+        ))
+    };
+
+    boxed_query
+        .execute(&mut conn)
+        .await
+        .and_then(|n| {
+            if n == 1 {
```
Comment (0x8f701) [inline src/actions.rs]: this should be `if n >= 1` it should be a mistake in `diesel-adapter` sorry...

File: src/actions.rs [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+            rule[1],
+            rule[2],
+            rule[3],
+            rule[4],
+            rule[5]
+        )
+        .execute(&mut transaction)
+        .await
+        .and_then(|n| {
+            if n == 1 {
```
Comment (0x8f701) [inline src/actions.rs:187]: this should be also `if n >= 1`.

File: src/actions.rs [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+        if n == 1 {
+            Ok(true)
+        } else {
+            Err(SqlxError::RowNotFound)
+        }
+    })
+    .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))
+}
+
+#[cfg(feature = "postgres")]
```
Comment (0x8f701) [inline src/actions.rs:108]: I think if the two functions are the same, we just write one and remove feature gate

File: src/actions.rs [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+            rule[1],
+            rule[2],
+            rule[3],
+            rule[4],
+            rule[5]
+        )
+        .execute(&mut transaction)
+        .await
+        .and_then(|n| {
+            if n == 1 {
```
Comment (0x8f701) [inline src/actions.rs:140]: this should be `if n >= 1`

File: src/actions.rs [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+            field_values[4],
+            field_values[5]
+        ))
+    };
+
+    boxed_query
+        .execute(&mut conn)
+        .await
+        .and_then(|n| {
+            if n == 1 {
```
Comment (0x8f701) [inline src/actions.rs]: this should be `if n >= 1`

File: src/actions.rs [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+        rule.ptype,
+        rule.v0,
+        rule.v1,
+        rule.v2,
+        rule.v3,
+        rule.v4,
+        rule.v5
+    )
+    .execute(&mut conn)
+    .await
```
Comment (0x8f701) [inline src/actions.rs:553]: we need to check if it has been added. `if n == 1`

File: src/actions.rs [hunk 12]
```diff
@@ -0,0 +1,621 @@
...
```
Comment (0x8f701) [inline src/actions.rs:529]: we need to check if it has been added. `if n == 1`

File: src/actions.rs [hunk 13]
```diff
@@ -0,0 +1,621 @@
...
+            rule.ptype,
+            rule.v0,
+            rule.v1,
+            rule.v2,
+            rule.v3,
+            rule.v4,
+            rule.v5
+        )
+        .execute(&mut transaction)
+        .await
```
Comment (0x8f701) [inline src/actions.rs:582]: we need to check if it has been added, `if n != 1` then we stop the loop and revert the transaction

File: src/actions.rs [hunk 14]
```diff
@@ -0,0 +1,621 @@
...
```
Comment (0x8f701) [inline src/actions.rs:621]: same as above we need to check if rule has been added

File: src/actions.rs [hunk 15]
```diff
@@ -0,0 +1,641 @@
...
+            rule[1],
+            rule[2],
+            rule[3],
+            rule[4],
+            rule[5]
+        )
+        .execute(&mut transaction)
+        .await
+        .and_then(|n| {
+            if n >= 1 {
```
Comment (0x8f701) [inline src/actions.rs]: here it should be `n==1` it seems

File: src/actions.rs [hunk 16]
```diff
@@ -0,0 +1,641 @@
...
```
Comment (0x8f701) [inline src/actions.rs]: here it should be `n==1` it seems, I thought you insert all at one time

File: src/actions.rs (hunk 15 above)
Comment (hackerchai) [reply src/actions.rs]: Yes, it is my mistake. Sorry

File: src/actions.rs (hunk 16 above)
Comment (hackerchai) [reply src/actions.rs]: same as above. fixed.

**Pull request merged**